kimchi/src/circuits/argument.rs

```diff
@@ -0,0 +1,58 @@
+//! An argument is simply a number of constraints,
+//! which we want to enforce on all points of the domain.
+//! Both the permutation and the plookup arguments fit this type.
+//! Gates can be seen as filtered arguments,
+//! which apply only in some points (rows) of the domain.
+
+use crate::{alphas::Alphas, circuits::expr::prologue::*};
+use ark_ff::FftField;
+use serde::{Deserialize, Serialize};
+
+use super::gate::GateType;
+
+/// A constraint type represents a polynomial that will be part of the final equation f (the circuit equation)
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Debug, Serialize, Deserialize)]
+pub enum ArgumentType {
+    /// Gates in the PLONK constraint system.
+    /// As gates are mutually exclusive (a single gate is set per row),
+    /// we can reuse the same powers of alpha across gates.
+    Gate(GateType),
+    /// The permutation argument
+    Permutation,
+    /// The lookup argument
+    Lookup,
+}
+
+/// The interface for a minimal argument implementation.
+pub trait Argument {
+    /// The field being used for this argument.
+    type Field: FftField;
+
+    /// The type of constraints that this will produce.
+    /// This is important to enforce that we don't combine the constraints
+    /// with powers of alpha that collide with other mutually inclusive arguments.
+    const ARGUMENT_TYPE: ArgumentType;
+
+    /// The number of constraints created by the argument.
+    const CONSTRAINTS: usize;
+
+    /// Returns the set of constraints required to prove this argument.
+    // TODO: return a [_; Self::CONSTRAINTS] once generic consts are stable
+    fn constraints(&self) -> Vec<E<Self::Field>>;
```

Comment:
I need `self` for the plookup argument

Reply:
thinking a bit more about that, we could potentially augment the interface like this:

```rust
trait Argument {
  type Input;
  fn constraints(input: Input) -> ...
}
```

what do you think about experimenting with these in another PRs though? Same for the other comment about hiding the ugliness :o

Reply:
This PR is an experiment itself. I'd prefer if we limited ourselves to just the gates for this PR, since they can be done cleanly. Perhaps there's then a nice abstraction for the non-local constraints, but it need not be the same one IMO

Reply:
I'm going to argue that we should get this in as this is much cleaner for the permutation and the lookup as well. It shows exactly where we create the constraints and it allows us to use the `Alphas` quite easily with it as well.

Reply:
I agree with David on the cleaner aspect of this approach. I am not saying this is not subject to further improvements, but so far I like the idea of having this `ArgumentType` for the three possible variants. We can think of ways to hide some ugliness in a future PR. And the same goes for a slightly different treatment of the non local constraints. But accepting this PR only for the gates would become a bit tricky, as we would be using quite different approaches for the other types of arguments, which I think could make the code a bit less readable.